### PR TITLE
Notebook on accelerating deep neural networks with tensorly

### PR DESCRIPTION
**Pull request by jacobgil**
A notebook on accelerating deep neural networks with TensorLy and PyTorch

### CODE REVIEW

File: 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb [hunk 1]
```diff
@@ -0,0 +1,594 @@
...
+    "\n",
+    "In this notebook we will show how to use TensorLy for accelerating convolutional layers in PyTorch.\n",
+    "\n",
+    "This notebook is based on this blog post: https://jacobgil.github.io/deeplearning/tensor-decompositions-deep-learning\n",
+    "\n",
+    "We will use the Tucker and CP decomositions to make low rank approximations of convolutional layers, and make them:\n",
+    "1. Faster.\n",
+    "2. Smaller.\n",
+    "\n",
+    "Like pruning, these are tools that can be used to make the networks smaller and faster, and then try to restore accuracy by training further.\n",
```
Comment (JeanKossaifi) [inline 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb]: Typo: (decomositions instead of decomposition) 
Also the two sentences are slightly redundant, we could replace by a similar paragraph such as:

We will show how to use Tucker and CP decompositions to find low rank approximations of convolutional layers.
Akin to prunning, this will allow use to reduce the number of parameters as well as make them faster. While the low-rank approximation can decrease the accuracy, this can be restored by training further.

File: 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+    "\n",
+    "---\n",
+    "\n",
+    "First lets get VGG19 from torch vision: "
```
Comment (JeanKossaifi) [inline 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb]: Let's load a pretrained VGG-19 network, using torch vision?
Also, it might be clearer to replace the `print(model)` by a short text explaining how the VGG net works (e.g. buch of conv, relu, pooling followed by two large FC layers before the final classification layer)

File: 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+    "The idea here is from [1412.6553 Speeding-up Convolutional Neural Networks Using Fine-tuned CP-Decomposition](https://arxiv.org/abs/1412.6553)\n",
+    "\n",
+    "We can approximate our 4 dimensional convolutional kernel with a CP decomposition of rank R:\n",
+    "\n",
+    "$ K(i, j, s, t) = \\sum_{r=1}^R K^x_r(i)K^y_r(j)K^s_r(s)K^t_r(t) $.\n",
+    "\n",
+    "Plugging this into the formula for the convolutional layer output form above:\n",
+    "\n",
+    "$ V(x, y, t) = \\sum_r\\sum_i \\sum_j \\sum_sK^x_r(i)K^y_r(i)K^s_r(s)K^t_r(t)X(x-i, y-j, s) $\n",
+    "$ = \\sum_rK^t_r(t) \\sum_i \\sum_j K^x_r(i)K^y_r(j)\\sum_sK^s_r(s)X(x-i, y-j, s) $ \n",
```
Comment (JeanKossaifi) [inline 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb]: I guess later on we can add some images and equivalent formulation to make the whole formula more human friendly :)

File: 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+      "('Ratio', 0.3634896334389113)\n"
+     ]
+    }
+   ],
+   "source": [
+    "from functools import reduce # Valid in Python 2.6+, required in Python 3\n",
+    "import operator\n",
+    "\n",
+    "def count_params(model):    \n",
+    "    return  sum(reduce(operator.mul, p.size(), 1) for p in model.parameters())\n",
```
Comment (JeanKossaifi) [inline 05_pytorch_backend/cnn_acceleration_tensorly_and_pytorch.ipynb]: since we use numpy already, we could use `np.sum` instead of `sum` and `np.prod` instead of the reduce on mul. This saves us from importing reduce and operator and might be slightly clearer to read :)

**Comment by jacobgil** [comment]
Thanks,
I will address the comments, move it to the PyTorch backend and add an example of running on an image.
Maybe after that we can merge as a first version.

I was also thinking about adding code for demonstrating fine tuning (like in my original repo that does that on a cat/dog classifier) to restore accuracy, 
But that means a lot of more additional code will need to be imported to the notebook, and I thought maybe it would become less clear. 
I wanted to avoid a situation where the actual code that does the decomposition would become the minority :-)

**Comment by jacobgil** [comment]
@JeanKossaifi 
- I addressed the comments.
- Switched to the PyTorch backend (and tested with Python 3).
- Forward pass on an image before and after the decomposition, to show that the network is still sensible.

I guess next big potential changes would be:
- Adding auto rank selection with VBMF.
- Fine tuning.
But i'm not sure if it will over complicate the notebook.

**Comment by jacobgil** [comment]
@JeanKossaifi 
Let me know if you think something should be changed for a merge.

**Comment by JeanKossaifi** [comment]
Awesome, thanks @jacobgil!

I agree, let's merge this and we can think on adding more / having a follow-up notebook! :)
